### PR TITLE
Setting non-string Label text

### LINKED ISSUE #2140: Setting Label text to a non-string should throw a meaningful error
This works
```typescript
var label = new Plottable.Components.Label("1");
```

This does not
```typescript
var label = new Plottable.Components.Label(1);
```

TODO: Please document with current error message. 

Typescript disallows that indeed, but we might make it more friendly for our JS users.

**Edit** After debating, we decided that this will not be solved by stringifying the number, but instead throw a meaningful message. More information on #2442

### PR DESCRIPTION
**Pull request by estas**
Add overloads for Label.text(), TitleLabel constructor and AxisLabel constructor.

Fixes #2140

### CODE REVIEW

File: src/components/label.ts [hunk 1]
```diff
@@ -54,14 +54,19 @@ export module Components {
...
      * Sets the Label's text.
      *
+     * @param {nubmer} displayText
      * @param {string} displayText
      * @returns {Label} The calling Label.
      */
+    public text(displayText: number): Label;
     public text(displayText: string): Label;
-    public text(displayText?: string): any {
+    public text(displayText?): any {
```
Comment (aicioara) [inline src/components/label.ts]: Use union instead
```typescript
public text(displayText: number | string): Label
```

File: src/components/label.ts [hunk 2]
```diff
@@ -54,14 +54,19 @@ export module Components {
     /**
      * Sets the Label's text.
      *
+     * @param {nubmer} displayText
      * @param {string} displayText
      * @returns {Label} The calling Label.
      */
+    public text(displayText: number): Label;
     public text(displayText: string): Label;
-    public text(displayText?: string): any {
```
Comment (aicioara) [inline src/components/label.ts]: Use a union instead
```typescript 
public text(displayText: number | string): Label
```

File: src/components/label.ts [hunk 3]
```diff
@@ -54,14 +54,19 @@ export module Components {
     /**
      * Sets the Label's text.
      *
+     * @param {nubmer} displayText
```
Comment (aicioara) [inline src/components/label.ts]: Amend

File: src/components/label.ts (hunk 1 above)
Comment (aicioara) [inline src/components/label.ts]: Amend the constructor as well

File: src/components/label.ts [hunk 4]
```diff
@@ -54,14 +54,19 @@ export module Components {
...
      * @returns {Label} The calling Label.
      */
+    public text(displayText: number): Label;
     public text(displayText: string): Label;
-    public text(displayText?: string): any {
+    public text(displayText?): any {
       if (displayText === undefined) {
         return this._text;
       } else {
+        if (typeof displayText == "number") {
```
Comment (aicioara) [inline src/components/label.ts]: Use `===`

Comment (aicioara) [inline src/components/label.ts]: Let's make this more future proof and not explicitly check if it is a number. Just cast it to a String. String casted to strings are fine.

File: src/components/label.ts [hunk 5]
```diff
@@ -54,14 +54,19 @@ export module Components {
...
+          displayText = displayText.toString();
```
Comment (aicioara) [inline src/components/label.ts]: Either use `"" + displayText` or catch the null case.
For catching null, just change line 64 to 
```typescript
if (displayText == undefined) {
```

File: src/components/label.ts [hunk 6]
```diff
@@ -157,10 +162,13 @@ export module Components {
     public static TITLE_LABEL_CLASS = "title-label";
     /**
      * @constructor
+     * @param {number} [text]
```
Comment (aicioara) [inline src/components/label.ts]: Same

File: src/components/label.ts [hunk 7]
```diff
@@ -157,10 +162,13 @@ export module Components {
...
      * @param {string} [text]
      * @param {number} [angle] One of -90/0/90. 0 is horizontal.
      */
-    constructor(text?: string, angle?: number) {
+    constructor(text: number, angle: number); 
+    constructor(text: string, angle: number);
+    constructor(text?, angle?) {
```
Comment (aicioara) [inline src/components/label.ts]: Same

File: src/components/label.ts [hunk 8]
```diff
@@ -170,10 +178,13 @@ export module Components {
     public static AXIS_LABEL_CLASS = "axis-label";
     /**
      * @constructor
+     * @param {number} [text]
```
Comment (aicioara) [inline src/components/label.ts]: Same

File: src/components/label.ts [hunk 9]
```diff
@@ -170,10 +178,13 @@ export module Components {
...
      * @param {string} [text]
      * @param {number} [angle] One of -90/0/90. 0 is horizontal.
      */
-    constructor(text?: string, angle?: number) {
+    constructor(text: number, angle: number); 
+    constructor(text: string, angle: number);
+    constructor(text?, angle?) {
```
Comment (aicioara) [inline src/components/label.ts]: Same

File: test/components/labelTests.ts [hunk 10]
```diff
@@ -189,4 +189,22 @@ describe("Labels", () => {
...
+  });
+
+  it("AxisLabel can takes number as displayText", () => {
+    var svg = TestMethods.generateSVG(100, 100);
+    var label = new Plottable.Components.AxisLabel(1);
+    label.renderTo(svg);
+    var text = (<any> label)._content.select("text");
+    assert.strictEqual(text.text(), "1", "text equal to input number");
+    svg.remove();
+  });
```
Comment (aicioara) [inline test/components/labelTests.ts]: A bit of code duplication here. Let's just test the `Label`. `TitleLabel` and `AxisLabel` should behave the same if `Label` works, as they do not override the method

File: test/components/labelTests.ts [hunk 11]
```diff
@@ -189,4 +189,22 @@ describe("Labels", () => {
     var testLabel = new Plottable.Components.Label("testing label");
     assert.throws(() => testLabel.padding(-10), Error, "Cannot be less than 0");
   });
+
+  it("TitleLabel can takes number as displayText", () => {
+    var svg = TestMethods.generateSVG(100, 100);
+    var label = new Plottable.Components.TitleLabel(1);
+    label.renderTo(svg);
+    var text = (<any> label)._content.select("text");
```
Comment (aicioara) [inline test/components/labelTests.ts]: `.content()` and no `<any>` casting

File: test/components/labelTests.ts [hunk 12]
```diff
@@ -189,4 +189,22 @@ describe("Labels", () => {
     var testLabel = new Plottable.Components.Label("testing label");
     assert.throws(() => testLabel.padding(-10), Error, "Cannot be less than 0");
   });
+
+  it("TitleLabel can takes number as displayText", () => {
+    var svg = TestMethods.generateSVG(100, 100);
+    var label = new Plottable.Components.TitleLabel(1);
```
Comment (aicioara) [inline test/components/labelTests.ts]: Wrap it inside an `assert.doesNotThrow()`

File: test/components/labelTests.ts [hunk 13]
```diff
@@ -189,4 +189,22 @@ describe("Labels", () => {
...
+    label.renderTo(svg);
```
Comment (aicioara) [inline test/components/labelTests.ts]: No need to render, this is expensive. Just check for `label.text()`

**Comment by aicioara** [comment]
Done with CR. 
Run 
```
grunt tslint
```
To check for style issues.

**Comment by estas** [comment]
@aicioara, thanks for code review! tslint said that all is ok.

File: src/components/label.ts [hunk 14]
```diff
@@ -15,10 +15,10 @@ export module Components {
      * A Label is a Component that displays a single line of text.
      *
      * @constructor
-     * @param {string} [displayText=""] The text of the Label.
+     * @param {number|string} [displayText=""] The text of the Label.
      * @param {number} [angle=0] The angle of the Label in degrees (-90/0/90). 0 is horizontal.
      */
-    constructor(displayText = "", angle = 0) {
+    constructor(displayText: number|string = "", angle: number = 0) {
```
Comment (aicioara) [inline src/components/label.ts]: explicitly typing angle as number is not needed

File: src/components/label.ts [hunk 15]
```diff
@@ -54,15 +54,15 @@ export module Components {
     /**
      * Sets the Label's text.
      *
-     * @param {string} displayText
+     * @param {nubmer|string} displayText
      * @returns {Label} The calling Label.
      */
-    public text(displayText: string): Label;
-    public text(displayText?: string): any {
+    public text(displayText: number|string): Label;
```
Comment (aicioara) [inline src/components/label.ts]: space before and after |
(we should probably have a TSLint rule for this, sorry)

File: src/components/label.ts [hunk 16]
```diff
@@ -54,15 +54,15 @@ export module Components {
     /**
      * Sets the Label's text.
      *
-     * @param {string} displayText
+     * @param {nubmer|string} displayText
```
Comment (aicioara) [inline src/components/label.ts]: typo

File: test/components/labelTests.ts [hunk 17]
```diff
@@ -189,4 +189,11 @@ describe("Labels", () => {
     var testLabel = new Plottable.Components.Label("testing label");
     assert.throws(() => testLabel.padding(-10), Error, "Cannot be less than 0");
   });
+
+  it("Label can takes number as displayText", () => {
+    assert.doesNotThrow(() => {
+      var label = new Plottable.Components.Label(1);
+      assert.strictEqual(label.text(), "1", "text equal to input number");
```
Comment (aicioara) [inline test/components/labelTests.ts]: Take this out of the assert. Extract `var label` as well for this to work.

**Comment by aicioara** [comment]
Done with CR2

File: test/components/labelTests.ts (hunk 17 above)
Comment (estas) [inline test/components/labelTests.ts]: Do you mean this? -
```javascript
it("Label can takes number as displayText", () => {
    assert.doesNotThrow(() => {
        (new Plottable.Components.Label(1)).text();
    })
})
```

Comment (aicioara) [inline test/components/labelTests.ts]:
```typescript 
  it("Label can takes number as displayText", () => {
    var label: Plottable.Components.Label;
    assert.doesNotThrow(() => {
      label = new Plottable.Components.Label(1);
    }, Error, "Label does not crash on invalid values");
    assert.strictEqual(label.text(), "1", "text equal to input number");
  });
```

File: test/components/labelTests.ts [hunk 18]
```diff
@@ -189,4 +189,11 @@ describe("Labels", () => {
...
+    }, Error, "Label does not crash on invalid values");
```
Comment (aicioara) [inline test/components/labelTests.ts]: This is not an invalid value anymore, so the message should say "Label does not crash on numeric values"

File: test/components/labelTests.ts (hunk 17 above)
Comment (estas) [inline test/components/labelTests.ts]: Oh, thanks.

**Comment by aicioara** [comment]
Looks great! Thank you. Sending it to QE.

@estas in addition, could you fill out this [Contributor License Agreement](https://github.com/palantir/plottable/blob/develop/Plottable_Personal_Contributer_License_Agreement.pdf?raw=true) and send it to opensource@palantir.com?

**Comment by estas** [comment]
No problem :)
I sent it yesterday.

**Comment by aicioara** [comment]
Thank you

**Comment by jtlan** [comment]
@aicioara is this ready to go?

**Comment by aicioara** [comment]
Got delayed a bit because of hack week.

**Comment by crmorford** [comment]
Functionally, this looks fine http://jsfiddle.net/1h335t30/1/. 
@estas, Sorry for keeping this pull request in limbo! This pull request piqued some discussions about data sanitization, and what we want to set precedents for moving forward. We're trying to make sure this is completely in line with what we decide before we merge it :)

**Comment by crmorford** [comment]
Hi @estas! 
The Plottable team has made the high level decision that users should be doing their own data sanitization.  This is because Plottable doesn't want to be responsible for assumptions about what pieces of data mean, and potentially make a wrong guess, thus causing unexpected behavior. Although coercing the text for labels is fairly innocent, we want to make sure we are consistent across the board.

What this means is that despite the fact that this pull request is otherwise ready to merge, unfortunately we're going to have to decline it.

Thank you so much for using and contributing to Plottable, and we want to encourage you to keep submitting pull requests. The team is going to go through our issues and remove any that are no longer relevant, so that this won't happen again :)

Let us know if you have any questions, we're happy to answer them anytime!

**Comment by aicioara** [comment]
@crmorford Just double checking, the outcome was that we want hard errors on such cases?

**Comment by bluong** [comment]
Yes

**Pull request closed without merging**